### ISSUE TITLE
Unable to get DefinePlugin, CommonsChunkPlugin and UglifyJsPlugin to work together.

### ISSUE THREAD

**Issue by richtera** [webpack/webpack#6328]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
When using DefinePlugin and UglifyJsPlugin together the minification is done before the DefinePlugin searches for variables. So things like process.env.NODE_ENV is not substituted inside of a CommonsChunkPlugin chunk. I have tried many combinations and disabling UglifyJSPlugin by removing it will correctly substitute defined variables in the chunk. So the UglifyJsPlugin somehow doesn't run last.

**If the current behavior is a bug, please provide the steps to reproduce.**

The main problem can be seen by including React using "import React from 'react';". This includes the index.js file inside of react and then should import either react.development.js or react.production.min.js depending on process.env.NODE_ENV. Without uglify I see if ("production" == "production") which can then get eliminated. But with Uglify I see if ("producation" == e.env.NODE_ENV) and env is always an empty object since the DefinePlugin doesn't create globals. I am not quite sure how to solve this.

I can put together a repo but I am trying to deploy a release and I am hoping someone else has already run into this.

**What is the expected behavior?**
The Uglify plugin should always run after the DefinePlugin has resolved globals in the js files.

**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

**Comment by ooflorent** [comment]
> **Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

Could you provide these informations please?

I did my best to reproduce the issue but everything works as expected. Maybe you are using an outdated version of webpack or did something wrong within your configuration (hard to tell without the full picture).

**Comment by richtera** [comment]
I am using Mac OS X 10.13.2 (17C205) running 8.9.1 of node.
webpack is webpack@3.10.0. My webpack config is definitely complicated and uses a Promise to return the config. Is there a way to see the sequence of webpack plugins running during packing?
I decreased the size of the project by removing all components and dependent files and then it works correctly with and without uglify. I seem to be hitting some kind of limit. I kept the same project structure and config but removed everything except for the "import React from 'react';" in each of the index.js files and it did correctly pull the production react code and eliminate the development require. But my normal build process with all files doesn't. Not sure how to analyze the failure.

**Comment by evilebottnawi** [comment]
@richtera can you create minimum reproducible test repo?

**Comment by richtera** [comment]
I can create a minimum repo but it works in the minimal configuration. I can't quite figure out what makes it break.

**Comment by ooflorent** [comment]
Could you post the resolved config? Even if it doesn't work, maybe we could find what's wrong

**Comment by richtera** [comment]
Here is a simplified version of the project.
https://github.com/richtera/webpack-project-for-debugging
It uses yarn to pull in the correct dependencies but this sample project only imports react in each of the index.js files.
NOTE: We use NODE_ENV=production to turn on and off the Uglify project.

**Comment by ooflorent** [comment]
I can't reproduce the issue using your repository. Here is what I did:

```
git clone git@github.com:richtera/webpack-project-for-debugging.git issue-6328
cd issue-6328
echo '{}' > VERSION.json
yarn install
NODE_ENV=production yarn build
```

Inspecting `commons.*.js` reveals the production build on React, as expected.
Without specifying `NODE_ENV`, the bundle contains:

```js
if (false) {
  module.exports = require('./cjs/react.production.min.js');
// not included, as expected
} else {
  module.exports = __webpack_require__(5);
}
```

Am I missing something?

**Comment by richtera** [comment]
That's correct without uglify and it's working in the repo without the code. Once I put the rest of my code in it breaks. Seems to be some kind of limit. When I look at my built version it looks like this:
```javascript
 * react.development.js
 *
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"production"!==e.env.NODE_ENV&&function(){function e(t){if(n
.
.
.
```

**Comment by ooflorent** [comment]
Sorry I can't reproduce. I used uglify and it worked as expected. What command did you use to build the project?

```sh
NODE_ENV=production yarn build              # uglify-ed build
grep -l NODE_ENV dist/assets/js/*.js        # nothing
grep -l '"production"' dist/assets/js/*.js  # nothing
```

Even without minification:

```sh
yarn build                                  # dev build
grep -l NODE_ENV dist/assets/js/*.js        # nothing
grep -l '"production"' dist/assets/js/*.js  # nothing
```

I think the problem is elsewhere. Your repro-repository works as expected. Without issue.

**Comment by richtera** [comment]
That's exactly my problem. My full project (using the same webpack config) fails but without all the source code it works normally. I don't know how to figure out what about the project source code causes uglify to fail in production.

**Comment by richtera** [comment]
Ok I was able to add a few files from my deployment to reproduce the error in the sample repo.

**Comment by richtera** [comment]
So far I narrowed it down to the CommonChunksPlugin. If I add some shared code and then set the minChunks to 2 I can make the error happen easily.

**Comment by richtera** [comment]
Does anyone know how to make sure to run the DefinePlugin before the UglifyPlugin in the case of a common chunk?

**Comment by richtera** [comment]
There might be a problem with just the DefinePlugin as well. When disabling Uglify it does seem to solve some of the instances of process.env.NODE_ENV, but I still find more of them in the output text. So there might be a combination of things going on. Is there a maximum limit of js code?

**Comment by richtera** [comment]
The latest test code in here uses a git submodule. Sorry I forgot to mention this.

**Comment by richtera** [comment]
I simplified the test project and added a README. It shows that the problem shows up with and without UglifyJsPlugin enabled. So I am suspecting the problem is the DefinePlugin rather than the UglifyJsPlugin.

**Comment by ottoo** [comment]
For me everything also works correctly with this plugin. I have defined env vars in a separate `env` variable and then have the NODE_ENV accordingly and its value is production as intended 

```
new webpack.DefinePlugin({
      env: envVars,
      'process.env.NODE_ENV': JSON.stringify('production')
    })
```

and 

```
new UglifyJsPlugin({
      uglifyOptions: {
        ie8: false
      },
      parallel: true
    })
```

**Comment by richtera** [comment]
Turns out of all the modules I am using those two are the only ones without node:false or other ways of building the output file. I posted two pull requested to fix those two modules.

- fakiolinho/react-loading#42
- borisyankov/react-sparklines#108

You can install temporarily directly from github.

- npm install or yarn add https://github.com/richtera/react-loading.git
- npm install or yarn add https://github.com/richtera/react-sparklines.git

**Comment by richtera** [comment]
Maybe I didn't explain this well enough. Both of the modules react-sparklines and react-loading contain the process polyfill in their main file.
Example from react-loading

```javascript
.
.
.
/******/ ([
/* 0 */
/***/ (function(module, exports) {

// shim for using process in browser
var process = module.exports = {};

// cached from whatever global is present so that test runners that stub it
// don't break things.  But we need to wrap it in a try catch in case it is
// wrapped in strict mode code which doesn't define any globals.  It's inside a
// function because try/catches deoptimize in certain engines.

var cachedSetTimeout;
var cachedClearTimeout;

function defaultSetTimout() {
    throw new Error('setTimeout has not been defined');
}
function defaultClearTimeout () {
    throw new Error('clearTimeout has not been defined');
}
(function () {
    try {
        if (typeof setTimeout === 'function') {
.
.
.
```

Any file included after that one in the bundle will get process injected instead of keeping the global.
After commenting out the dependencies and usage of those two modules webpack does the correct DefinePlugin substitutions.

I am not sure how many of these kinds of modules are around that don't have node: false in their webpack config.

**Comment by ooflorent** [comment]
@richtera thanks for the explanation. Unfortunately webpack can’t do anything about this. You need escalate the issue those vendors bug trackers.

**Comment by richtera** [comment]
It would be nice if the output log showed whether node support was used. Also maybe DefinePlugin should not break with a process shim that's part of webpack or at least shows a warning if it seems process being injected.

**Comment by ooflorent** [comment]
It’s more tedious than this.

**tl;dr** webpack can’t know about the shim.

If a code using `process` is bundled, then the bundled will add `var process = require("process");` in the file. After this operation, the `process` binding no longer reference the global `process` binding.

Moreover, it is most probably not a `require` call but a call to `__webpack_require__` or browserify own `require`. Webpack can’t handle this because it would break this:

```js
var process = someFunc();
```

**Comment by ooflorent** [comment]
Same apply for all shimmed global bindings.

**Comment by richtera** [comment]
I am not suggesting "fixing it" I am suggesting to output a list of injected variables and if those variables contain something DebugPlugin replacements start with output a warning. I agree that this cannot automatically resolve the problem but it was really hard to find and resolve this problem by folding and grepping the output file. There has to be a way to output the variables that are being injected as "globals".

**Comment by ooflorent** [comment]
> I am suggesting to output a list of injected variables and if those variables contain something DebugPlugin replacements start with output a warning.

That's possible to be done. However, there as multiple internal plugins that rely on globals (`process`, `module`). For example, webpack's source contains at least 500 overridden bindings (based on a regex search). What I'm trying to say is that overriding bindings is quite common.

Creating a plugin that _flag_ redefinitions is a good idea. If you want, you could create PR on `next`branch and we would happily review it! That would definitely be a good addition in development mode.

**Comment by richtera** [comment]
Hmmm, ok. Other question then. When process is being injected, would there be a way to still make DefinePlugin work? That might actually be a more reliable approach.

**Comment by ooflorent** [comment]
Unfortunately not. Ignoring binding redefinitions would be a breaking change. The issue is that vendors should not include the shim. It’s not webpack’s responsibility to fix their mistakes.

**Comment by richtera** [comment]
But if there was a reason to include the process shim for some other reason should the user expect that process.env.NODE_ENV replacement no longer works? Seems and unintended side effect if node process is required. But maybe there is no way. I guess we'll have to put some script to make sure the build worked.

```sh
if grep 'env.NODE_ENV' dist/assets/js/*.js
then
   echo "Build failed."
   exit 1
fi
```

**Comment by richtera** [comment]
Maybe we should have a "find broken modules" plugin which specifically searches for globals that DefinePlugin might want to replace.

**Comment by richtera** [comment]
I can't figure out how webpack detects the fact that a module returns process in order to inject it into the next modules that reference process. Would it be possible to disable that detection in addition to the automatic injection of process? Because if a module includes a process shim and only uses it internally I don't think that would be a problem.

**Comment by ooflorent** [comment]
Webpack cannot directly operates on the source code. It needs to _parse_ it into an [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree). This data structure describes what the file is made of. It is a mandatory step in order to perform a static analysis of your program.

While _traversing_ the AST, webpack tracks what is in the scope. For example, a variable declaration adds a new binding into the scope, whereas a function declaration also adds a binding but creates a child scope. These hierarchical scopes contain all bindings and renames performed inside the file.

Let's consider the following file:

```js
const modules = new Map()

module.exports = {
  registerModule(module) {
    modules.set(module.id, module)
  }
}
```

`module` references the module being evaluated but also a function argument. Webpack must do scope analysis in order to _know_ what `module` is. Moreover, `module.id` is evaluated at compile-time so replacing it in the above function is wrong. **That's the reason why there is scope-analysis.**

**What about `DefinePlugin`?**

This plugins accepts an object which contains the _bindings_ to inline at compile-time. It is commonly used to inline `process.env.NODE_ENV` but you could specify all the bindings you can think about. For example you can _unbind_ all Harmony global bindings (note: don't do that):

```js
new DefinePlugin({
  Symbol: "undefined",
  Promise: "undefined",
  Map: "undefined",
  Set: "undefined",
  WeakMap: "undefined",
  WeakSet: "undefined",
})
```

If you add such plugin to your build, you will have to define them before using them:

```js
/* example 1 */
new Map() // throws because webpack will turn it into `new undefined()`

/* example 2 */
const Map = require("es6-map")
new Map() // ok
```

**So, how is evaluated `process.env.NODE_ENV`?**

`DefinePlugin` splits the pattern into multiple parts and checks if it is still safe to perform the replacement. If not, webpack won't do it. **The replacement only occurs on global bindings.** In most case, there is no issue. However, if the first part of the pattern (here `process`) is no longer what is expected (a global binding named `process`), the replacement is skipped. That's why if you consume a bundled package where all occurrence of `process` are file scoped bindings instead of global bindings, nothing happen.

I admit, finding `env.NODE_ENV` isn't hard but `DefinePlugin` does not operate only on `process`. It analyzes every binding that it is fed of!

**Comment by kzc** [comment]
I think `webpack`'s inclusion of the node browser polyfill by default into bundles that happen to reference a node global is an undesirable feature. The majority of browser apps and libraries do not want that polyfill by default. They are just dealing with the `process.env.NODE_ENV` development/production legacy. The node browser polyfill should be an explicit opt-in to avoid very difficult to diagnose problems like this.
